collections/collections.go

```diff
@@ -0,0 +1,81 @@
+package collections
+
+import (
+	"errors"
+	"math"
+
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+)
+
+// ErrNotFound is returned when the provided key is not present in the StorageProvider.
+var ErrNotFound = errors.New("collections: not found")
+
+// ErrEncoding is returned when something fails during key or value encoding/decoding.
+var ErrEncoding = errors.New("collections: encoding error")
+
+// StorageProvider represents sdk.Context
+// it is used to avoid to reduce dependencies.
+type StorageProvider interface {
+	// KVStore returns a KVStore given its StoreKey.
+	KVStore(key storetypes.StoreKey) storetypes.KVStore
+}
+
+// Prefix defines a segregation namespace
+// for specific collections objects.
+type Prefix struct {
+	raw []byte // TODO(testinginprod): maybe add a humanized prefix?
+}
+
+// Bytes returns the raw Prefix bytes.
+func (n Prefix) Bytes() []byte { return n.raw }
+
+// NewPrefix returns a Prefix given the provided namespace identifier.
+// Prefixes of the same module must not start
+func NewPrefix[T interface{ int | string | []byte }](identifier T) Prefix {
+	i := any(identifier)
+	var prefix []byte
+	switch c := i.(type) {
+	case int:
+		if c > math.MaxUint8 || c < 0 {
+			panic("invalid integer prefix value: must be between 0 and 255")
+		}
+		prefix = []byte{uint8(c)}
+	case string:
+		prefix = []byte(c)
+	case []byte:
+		prefix = c // maybe copy?
+	}
+	return Prefix{raw: prefix}
+}
+
+// KeyEncoder defines a generic interface which is implemented
+// by types that are capable of encoding and decoding collections keys.
+type KeyEncoder[T any] interface {
+	// PutKey writes the key bytes into the buffer. Returns the number of
+	// bytes written.
+	// The bytes written must match the value returned by the Size method.
+	PutKey(buffer []byte, key T) (int, error)
```

Comment:
I think io.Reader/Writer makes sense too, but this API I think is okay. The documentation just needs to be a bit clearer and state that the buffer must be at least as large as the value returned by `Size`.

Reply:
An interface with methods that accept a byte slice which is expected to be of a certain minimum size defined by the return value of a different method of that interface is, at best, highly nonidiomatic. For those methods which accept byte slices to _panic_ when that expectation is not met is basically non-viable. Many issues here.

- First of all, because this isn't a reasonable invariant in the first place; there is no idiomatic Go API which takes a byte slice that is assumed to be of a dynamic size
- Second, because this invariant is dynamic based on implementation, which means failures are runtime errors rather than programmer errors, and (basically) not the kind of thing that should induce panics in the first place
- Third, because these methods return errors, and so any violation of this expectation is far more effectively communicated to callers that way
- Fourth, because the interface documents nothing re: concurrency, meaning there is no guarantee that the Size returned by an implementation of this interface at time t1 will be the same value that's expected by the e.g. PutKey method called at some future time t2, which makes it unsound at the level of the language memory model
- (several more things I elide for brevity)

Happy to explain any of these points in more detail, and/or to help improve on any of these points.

Reply:
@peterbourgon [Go's binary encoding API follows the same semantics](https://github.com/golang/go/blob/master/src/encoding/binary/binary.go#L38). Only difference being:
- For certain types, you have constant size (int64, uint64, ...) [perfectly inline with go's binary encoding API]
- For others types, you have variable size ([]byte, string).

> which means failures are runtime errors rather than programmer errors

A failure here might indicate a bad interface implementation, should I panic or not in this case? We can also error for instance by checking if the length matches the expected size.

There's also a second variant to this API which works with `append`.

`PutKey` would become: `AppendKey(buffer []byte, key K) (newBuffer []byte, error)`, would this satisfy your expectations?

Reply:
@testinginprod 

> Go's binary encoding API follows the same semantics

binary/encoding offers two "levels" of API. The lower level API you're pointing out here operates on byte slices directly and panics when invariants aren't met like e.g. the slices aren't big enough. The higher level API operates on io.Readers and io.Writers and doesn't panic.

> A failure of slice out of bounds kind indicates a bad interface implementation,

I'm assuming that "a failure of slice out of bounds" means "providing a slice to PutKey which is smaller than the value returned by Size", is that right? Assuming so, this has nothing to do with the type that implements the interface, right? The implementation defines the PutKey and Size methods, but it can't make any guarantees about anything between them. If an implementation defines Size to return 8, and PutKey is called with a buffer of size 4, this isn't anything to do with the interface implementation, it's a function of the caller.

> should I panic or not in this case? We can also error for instance by checking if the length matches the expected size.

Calling PutKey with a buffer slice that's too-small is a bug, but it's not a programmer error, and should definitely not panic. It should return an error. Better yet, the API shouldn't use byte slices at all, it should operate on io.Readers and io.Writers. They are in fact _faster_ than operating on byte slices in memory, in general; anyone claiming otherwise without concrete performance data is, bluntly, lying.

Reply:
> The implementation defines the PutKey and Size methods, but it can't make any guarantees about anything between them.

This is true but the consumer of the collectionsKeyEncoder API is the collections.Map itself, in its encodeKey function: https://github.com/cosmos/cosmos-sdk/pull/14134/files#diff-7ac7fb13c90a4739690a55f679a39d02a650034f043b44875b4e205ea20807d0R125-R129

If calling `StringKeyEncoder.Size("hello")` returns `5` and I initialise a 5-length bytes slice and pass it to `StringKeyEncoder.Encode(bufferOfLength5, "hello")` and it panics because of slice out of bounds then it is by fact a bad interface implementation, for instance `KeyEncoder` is still free to return an error in its `Encode` function, which can contain the buf length check, if desired.

> They are in fact faster than operating on byte slices in memory

TBH, I'd like the this to be proven, because I am not seeing how an `io.Writer/Reader` can give me any performance benefits, in `KeyEncoders` specifically, when I do have knowledge that: 
- Key byte size is ephemeral in 90% of cases 8-20-40-100 bytes, and I am exagerating.
- This key ends up in `KVStore` which does not accept `io.Reader/Writer` but works on `[]byte`.

For the `collections.Map` I'm not seeing how an abstraction over `[]byte` can be more performant than the raw `[]byte` itself, especially considering we don't have memory constraints in this context.

> Oh! Great! In that case, un-export the type and its methods, and all of my concerns go away 👍

This is unfortunately impossible because, for backwards compatibility reasons, I need to allow developers to implement their own `KeyEncoders` in order to retain their serialisation semantics around, let's say, `string` rather than using our own.

Reply:
> If calling StringKeyEncoder.Size("hello") returns 5 and I initialise a 5-length bytes slice and pass it to StringKeyEncoder.Encode(bufferOfLength5, "hello") and it panics because of slice out of bounds then it is by fact a bad interface implementation, also KeyEncoder is still free to return an error in its Encode function, which can contain the buf length check, if desired.

KeyEncoder is an interface which types can implement. And StringKeyEncoder is a type that would implement KeyEncoder, it isn't itself a value that callers would call methods on. So you wouldn't ever call StringKeyEncoder.Size, you'd always construct an enc = NewStringKeyEncoder, and then call enc.Size. 

But wait. You define Size(value T) int, and then you say that Size("hello") would return 5, implying e.g. Size("xxx") would return 3. That means Size isn't a function of the _type_, it's a function of the _value_. That would mean Encode(buf []byte, val T) would demand buf sizes which were not established by the type T itself, but rather by the specific val of type T. How can this possibly work? What do you expect your callers to do?

```go
enc := NewKeyEncoder[keytype, valtype](...)
key := keytype(...)
keysz := enc.Size(key)
buf := make([]byte, keysz)
err := enc.Encode(buf, key)
// use buf
```

This is clearly non-viable, right?